### PR TITLE
Kernel IR: Introducing kir::IrBuilder

### PR DESCRIPTION
**Pull request by tlemo**
This PR introduces a new interface for creating Kernel IR nodes: kir::IrBuilder. This is the only way to create new Kernel IR nodes,(so it's easy to track them), and it makes the connection between the IR nodes and the target kernel more explicit.

If the Kernel object is readily available, an IrBuilder can be "wrapped" around it directly:
`kir::IrBuilder ir_builder(kernel);`

During lowering, another option is to create an IrBuilder for the kernel that is being created:
`kir::IrBuilder ir_builder(GpuLower::current()->kernel());`

Once we have an IR builder instance, creating nodes looks

`auto new_node = ir_builder.create<kir::Int>(1));`
`auto result = ir_builder.mulExpr(lhs, rhs);`

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_ir.h [hunk 1]
```diff
@@ -22,12 +21,25 @@ namespace jit {
...
 
+class IrBuilder;
+
+//! Token used to restrict the access to Kernel IR constructors
+//!
+//! Granular "friendship" token, used to implement the "passkey" idiom:
+//! https://www.spiria.com/en/blog/desktop-software/passkey-idiom-and-better-friendship-c
+//! https://arne-mertz.de/2016/10/passkey-idiom
+//!
+class Passkey {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_ir.h:32]: Just to make sure I understand this correctly, this is to enforce that IrBuilder is the only interface for creating KIR nodes, isn't it?

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/kernel_ir.h:32]: Correct. We want to restrict the creation of IR nodes, so we don't have things like accidental `new kir::Int()` .

Private constructors + `friend` in the node themselves is an obvious alternative, although the passkey approach has a number of benefits:
1. It's granular - gives controlled access only to the constructors, not to everything private in the nodes
2. Preserves the logical "public" interface of the nodes (code readability)
3. It's more reliable than the raw C++ friend access (for ex. with passkey even the IR nodes can't create themselves w/o going through the IrBuilder interface)

**Review by naoyam — APPROVED** [review]
LGTM. Just left a small question. Thanks so much for refactoring the IR interface!

File: torch/csrc/jit/codegen/cuda/kernel_ir.h (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/kernel_ir.h:32]: Yeah, interesting pattern. I'm glad I learned something new.

File: torch/csrc/jit/codegen/cuda/kernel.h [hunk 2]
```diff
@@ -74,25 +84,40 @@ class TORCH_CUDA_API Kernel final : public NonCopyable {
...
-  // Lowered expressions
-  std::vector<Expr*> exprs_;
+  // Kernel IR nodes
+  std::vector<std::unique_ptr<Statement>> ir_nodes_;
+
+  // Map from value to its definition expression
+  std::unordered_map<const Val*, Expr*> definitions_;
+
+  // Top level expressions
+  std::vector<Expr*> top_level_exprs_;
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/kernel.h:120]: What do you mean by `top level expressions`?

**Review by rdspring1 — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/kernel.h (hunk 2 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/kernel.h:120]: This is the list of expressions at the top kernel scope, not all the expressions in the Kernel IR. These top level expressions may have arbitrary nested expressions (ex. for loops) - ie. the reason we need the recursive traversal in `LocalSyncInserter`.

Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/kernel.h:120]: Makes sense, since `exprs` implies all expressions in the kernel. 👍

**Pull request merged**